### ISSUE TITLE
Protocol Analyzer: doesn't list all USB 2.0 SETUP packets

### ISSUE THREAD

**Issue by dan-gies** [glscopeclient/scopehal#723]

**Comment by azonenberg** [comment]
Moving issue to scopehal as this is likely a back-end issue with the protocol decode class, not the GUI protocol analyzer.

**Comment by azonenberg** [comment]
Confirmed it decodes the same for me. The "unknown end PID" is likely something to do with it, but I won't know for sure until I dig deeper (at $dayjob right now so can't spend time on it).

**Comment by dan-gies** [comment]
No rush. Note that all 6 `SETUP` packets don't receive an `ACK` (which is the issue I'm diagnosing).

**Comment by azonenberg** [comment]
Yes. This might be a case of the decode correctly handling valid traffic but choking on out of spec data.

Generally I try to err on the side of "decode malformed data anyway, but color code or otherwise flag it as something being wrong". So there's probably room to improve.

**Comment by dan-gies** [comment]
That seems like a valid approach. The thing that's confusing me here is that the protocol analyzer shows _some_ `SETUP` packets and not others, when they are all identical.

**Comment by azonenberg** [comment]
+ @miek since you probably have more low level familiarity with USB than me. Is this traffic malformed or not?

**Comment by dan-gies** [comment]
I think it is, since the USB hub is failing to enumerate with the peripheral.

**Comment by miek** [comment]
> * @miek since you probably have more low level familiarity with USB than me. Is this traffic malformed or not?

Well.. yes and no. The device should be responding with either an ACK or NAK, but it's also a perfectly valid path through the state machine for the host to timeout waiting for a response and try again.

I've put a proposed fix in #741 that seems to work well:

![usb_setup_missing_ack](https://user-images.githubusercontent.com/578095/205682630-f9ba7f5a-0b5e-4242-b611-7f0fadd0a7ed.png)
